### PR TITLE
UOE-10388: Support for Applovin Max Response

### PR DESCRIPTION
**Pull request by AvinashKapre**

### CODE REVIEW

File: endpoints/openrtb2/auction.go [hunk 1]
```diff
@@ -382,6 +384,13 @@ func sendAuctionResponse(
 		}
 	}
 
+	//Handling for the Max Applovin Rejected case
+	rCtx := pubmatic.GetRequestCtx(ao.HookExecutionOutcome)
+	if rCtx != nil && rCtx.IsMaxRequest && response.ID == models.MaxRejected {
```
Comment (pm-nilesh-chate) [inline endpoints/openrtb2/auction.go:389]: Move this condition to function `ApplyMaxAppLovinResponseHeader()`

File: modules/pubmatic/openwrap/max.go [hunk 2]
```diff
@@ -219,3 +220,50 @@ func setIfKeysExists(source []byte, target []byte, keys ...string) []byte {
...
+	if bidResponse.NBR != nil && *bidResponse.NBR != 0 {
+		if !rctx.Debug {
+			*bidResponse = openrtb2.BidResponse{
+				ID: models.MaxRejected,
+			}
+		}
+		return bidResponse
+	}
+
+	respString, err := json.Marshal(bidResponse)
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/max.go:234]: nitpick: marshal returns byte so would suggest to rename `respString` to either `respBytes` or simply `resp`.

File: modules/pubmatic/openwrap/max.go [hunk 3]
```diff
@@ -219,3 +220,50 @@ func setIfKeysExists(source []byte, target []byte, keys ...string) []byte {
...
+	if err != nil {
+		*bidResponse = openrtb2.BidResponse{
+			ID: models.MaxRejected,
+		}
+		return bidResponse
+	}
+
+	signaldata := `{"signaldata":` + strconv.Quote(string(respString)) + `}`
+
+	if bidResponse != nil && len(bidResponse.SeatBid) > 0 && len(bidResponse.SeatBid[0].Bid) > 0 {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/max.go:244]: Above Marshal can be avoided with moving these check at start

File: modules/pubmatic/openwrap/targeting.go [hunk 4]
```diff
@@ -110,7 +110,7 @@ func (m OpenWrap) addPWTTargetingForBid(rctx models.RequestCtx, bidResponse *ope
 				if m.featureConfig.IsFscApplicable(rctx.PubID, seatBid.Seat, bidCtx.DspId) {
 					bidCtx.Fsc = 1
 				}
-			} else if !rctx.SendAllBids {
+			} else if !rctx.SendAllBids && !rctx.IsMaxRequest {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/targeting.go:113]: can we have `IsMaxRequestDebugEnabled()`

File: modules/pubmatic/openwrap/auctionresponsehook_test.go [hunk 5]
```diff
@@ -1935,3 +1936,114 @@ func TestOpenWrap_handleAuctionResponseHook(t *testing.T) {
...
+						},
+					},
+				},
+			},
+			want: want{
+				bidResponse: &openrtb2.BidResponse{
+					ID:    "123",
+					BidID: "456",
+					Cur:   "USD",
+					SeatBid: []openrtb2.SeatBid{
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/auctionresponsehook_test.go:2005]: Considering this as end-to-end UT, assert all the fields in response (i.e. fields returned by adapter but reset as part of maxapplovin resp) and include adm as well to assert injectTracker does not run?.

**Review by pm-nilesh-chate — CHANGES_REQUESTED** [review]

File: endpoints/openrtb2/auction_ow.go [hunk 6]
```diff
@@ -53,6 +54,11 @@ func UpdateResponseExtOW(bidResponse *openrtb2.BidResponse, ao analytics.Auction
 		return
 	}
 
+	if rCtx.IsMaxRequest {
+		upadteResponseExtForMax(ao, rCtx, bidResponse)
```
Comment (pm-saurabh-narkhede) [inline endpoints/openrtb2/auction_ow.go:58]: nitpick: should be updateResponseExtForMax

File: modules/pubmatic/openwrap/max.go [hunk 7]
```diff
@@ -219,3 +220,50 @@ func setIfKeysExists(source []byte, target []byte, keys ...string) []byte {
...
+					},
+				},
+			},
+		}
+	} else {
+		*bidResponse = openrtb2.BidResponse{
+			ID: models.MaxRejected,
+		}
+	}
+	return bidResponse
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/max.go:268]: can we keep debug in orignal response Ext?

**Review by pm-isha-bharti — APPROVED** [review]

File: modules/pubmatic/openwrap/applovinmax.go [hunk 8]
```diff
@@ -235,3 +235,58 @@ func updateAppLovinMaxRequest(requestBody []byte) []byte {
...
+		}
+		return maxAppLovin
+	}
+
+	if len(bidResponse.SeatBid) == 0 || len(bidResponse.SeatBid[0].Bid) == 0 {
+		maxAppLovin.Reject = true
+		return maxAppLovin
+	}
+	return maxAppLovin
+}
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/applovinmax.go:254]:
```
func updateMaxAppLovinResponse(rctx models.RequestCtx, bidResponse *openrtb2.BidResponse) models.MaxAppLovin {
	maxAppLovin := models.MaxAppLovin{Reject: false}

	if bidResponse.NBR != nil {
		if !rctx.Debug {
			maxAppLovin.Reject = true
		}
	} else if len(bidResponse.SeatBid) == 0 || len(bidResponse.SeatBid[0].Bid) == 0 {
		maxAppLovin.Reject = true
	}
	return maxAppLovin
}
```

File: modules/pubmatic/openwrap/applovinmax.go [hunk 9]
```diff
@@ -235,3 +235,54 @@ func updateAppLovinMaxRequest(requestBody []byte) []byte {
...
+
+func applyMaxAppLovinResponse(rctx models.RequestCtx, bidResponse *openrtb2.BidResponse) *openrtb2.BidResponse {
+	if rctx.AppLovinMax.Reject {
+		*bidResponse = openrtb2.BidResponse{}
+		return bidResponse
+	}
+
+	if bidResponse.NBR != nil {
+		return bidResponse
+	}
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/applovinmax.go:260]: add comment that this condition will be true only when debug is enabled

File: modules/pubmatic/openwrap/applovinmax_test.go [hunk 10]
```diff
@@ -675,3 +678,224 @@ func TestUpdateRequestWrapper(t *testing.T) {
...
+						},
+					},
+				},
+			},
+			want: models.AppLovinMax{
+				Reject: true,
+			},
+		},
+	}
+	for _, tt := range tests {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/applovinmax_test.go:757]: add case with "bidresponse w/o NBR and debug is disabled and seatbid.bid is valid"

File: modules/pubmatic/openwrap/applovinmax_test.go [hunk 11]
```diff
@@ -675,3 +678,224 @@ func TestUpdateRequestWrapper(t *testing.T) {
...
+					NBR: ptrutil.ToPtr(nbr.InvalidPlatform),
+				},
+			},
+			want: &openrtb2.BidResponse{
+				ID:  "123",
+				NBR: ptrutil.ToPtr(nbr.InvalidPlatform),
+			},
+		},
+		{
+			name: "failed to marshal bidresponse",
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/applovinmax_test.go:820]: This does not fail at unmarshal

File: modules/pubmatic/openwrap/applovinmax_test.go [hunk 12]
```diff
@@ -675,3 +678,224 @@ func TestUpdateRequestWrapper(t *testing.T) {
...
+						Bid: []openrtb2.Bid{
+							{
+								ID:    "456",
+								ImpID: "789",
+								Price: 1.0,
+								BURL:  "http://example.com",
+								Ext:   json.RawMessage(`{"signaldata":"{\"id\":\"123\",\"seatbid\":[{\"bid\":[{\"id\":\"456\",\"impid\":\"789\",\"price\":1,\"burl\":\"http://example.com\",\"adm\":\"\\u003cimg src=\\\"http://example.com\\\"\\u003e\\u003c/img\\u003e\",\"ext\":{\"key\":\"value\"}}],\"seat\":\"pubmatic\"}],\"cur\":\"USD\",\"ext\":{\"key\":\"value\"}}"}`),
+							},
+						},
+					},
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/applovinmax_test.go:890]: Should seat be required field?

File: modules/pubmatic/openwrap/targeting.go [hunk 13]
```diff
@@ -110,7 +110,7 @@ func (m OpenWrap) addPWTTargetingForBid(rctx models.RequestCtx, bidResponse *ope
 				if m.featureConfig.IsFscApplicable(rctx.PubID, seatBid.Seat, bidCtx.DspId) {
 					bidCtx.Fsc = 1
 				}
-			} else if !rctx.SendAllBids {
+			} else if !rctx.SendAllBids && isMaxRequestDebugEnabled(rctx) {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/targeting.go:113]: This should be OR

File: modules/pubmatic/openwrap/util.go [hunk 14]
```diff
@@ -342,3 +342,7 @@ func GetRequestUserAgent(body []byte, request *http.Request) string {
 	}
 	return request.Header.Get("User-Agent")
 }
+
+func isMaxRequestDebugEnabled(rctx models.RequestCtx) bool {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/util.go:346]: add UT

File: modules/pubmatic/openwrap/models/constants.go [hunk 15]
```diff
@@ -527,7 +527,8 @@ const (
 
 // constants for applovinmax requests
 const (
-	Agent        = "agent"
-	AppLovinMax  = "max"
-	TypeRewarded = "rewarded"
+	MaxRejected      = "max_rejected"
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/models/constants.go:530]: this is not needed

File: modules/pubmatic/openwrap/applovinmax.go [hunk 16]
```diff
@@ -235,3 +235,55 @@ func updateAppLovinMaxRequest(requestBody []byte) []byte {
...
+					{
+						ID:    bidResponse.SeatBid[0].Bid[0].ID,
+						ImpID: bidResponse.SeatBid[0].Bid[0].ImpID,
+						Price: bidResponse.SeatBid[0].Bid[0].Price,
+						BURL:  bidResponse.SeatBid[0].Bid[0].BURL,
+						Ext:   json.RawMessage(signaldata),
+					},
+				},
+			},
+		},
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/applovinmax.go:286]: Seat should be set

File: modules/pubmatic/openwrap/applovinmax_test.go (hunk 12 above)
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/applovinmax_test.go:919]: Need to check with OW SDK.

File: modules/pubmatic/openwrap/util.go (hunk 14 above)
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/util.go]: Removed function

File: modules/pubmatic/openwrap/applovinmax.go (hunk 16 above)
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/applovinmax.go:286]: Need to check with OW SDK

**Review by pm-saurabh-narkhede — APPROVED** [review]

**Review by pm-nilesh-chate — APPROVED** [review]

**Pull request merged**